### PR TITLE
fix: restart the sbt server when -D options change

### LINKED ISSUE #9682: [2.x] -D flags are silently ignored when an sbt server is already running
`sbt --help` says `-Dkey=val` passes the property to the java runtime. In sbt 2 the thin
client is the default, so the first invocation starts a background server and every later
one attaches to it. Any `-D` on those later invocations is silently droped. No warning,
exit code 0, and the build behaves as if the flag was never typed. This cost me an
afternoon on `-Dsbt.override.build.repos=true`.

With `sbt.version = 2.0.6` and a task that prints `sys.props.getOrElse("my.prop", "<unset>")`:

```
$ sbt -batch -Dmy.prop=first  showProp     # my.prop = first
$ sbt -batch -Dmy.prop=second showProp     # my.prop = first   <-- wrong
$ sbt shutdown
$ sbt -batch -Dmy.prop=third  showProp     # my.prop = third
```

The task needs `Def.uncached`, otherwise sbt replays the cached result and prints nothing,
which makes it even harder to spot.

I'd expect either a warning that the running server was started with different properties,
or a restart. Silence is the worst option, because nothing in the output distinguishes
"flag applied" from "flag thrown away". Forwarding them to a live server isn't really
possible, `sbt.override.build.repos` and friends are read by the launcher at JVM startup.

Workaround is `sbt shutdown` first, or `sbt --server`, which gets a fresh JVM every time.
The same applies to `.sbtopts`, since it feeds the same JVM arguments.

sbt 1 doesn't have this. Plain `sbt` was a fresh JVM per invocation, and `sbtn` had the
same limitation but was opt-in. #7663 is adjacent but about client flags like `--no-color`,
not boot properties.

sbt 2.0.6, runner script 2.0.7, launcher 2.0.7, Temurin 25.0.4.1, Linux.

### PR DESCRIPTION
**Pull request by stasimus**
`-D` options only reach a server the client starts itself, so once one was running they were dropped without a word (#9682). The client now records them in `active.json` and restarts the server when a later invocation differs. Skipped for bsp, completions and the JVM client. `-Dsbt.server.autorestart=false` opts out.

### CODE REVIEW

File: protocol/src/main/contraband/portfile.contra [hunk 1]
```diff
@@ -9,6 +9,8 @@ type PortFile {
   uri: String!
   tokenfilePath: String
   tokenfileUri: String
+  ## Newline separated -D options the thin client passed to this server when it started it.
+  sysProps: String @since("2.0.7")
```
Comment (eed3si9n) [inline protocol/src/main/contraband/portfile.contra]: I suggest using a list here, which would turn into a Vector in Scala.

```suggestion
  sysProps: [String] @since("2.0.7")
```

**Comment by stasimus** [comment]
Was a flaky test - I will try to address in separate PR

**Comment by stasimus** [comment]
Fix for test flaky test went in #9687, this PR is already bigger than I expected

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala [hunk 2]
```diff
@@ -1258,6 +1379,45 @@ object NetworkClient {
...
+    "sbt.script",
+    "sbt.server.autorestart",
+    "sbt.server.autostart",
+    "sbt.supershell",
+  )
+
+  /** The `-D` options that make up the identity of a server started with `sbtArguments`. */
+  private[client] def serverSysProps(sbtArguments: Seq[String]): Seq[String] =
+    sbtArguments
+      .filter(_.startsWith("-D"))
```
Comment (anatoliykmetyuk) [inline main-command/src/main/scala/sbt/internal/client/NetworkClient.scala:1470]: The props may come not only from `-D` command line args but also from other sources, such as: `.sbtopts`, `.jvmopts`, `SBT_OPTS`, or launcher options such as `--sbt-version` and `--java-home`.

Consider the following failure mode this PR does not account for:

Start with .sbtopts containing -Dmy.prop=first, change it to second, then run the property-printing task again; it still prints first without restarting.

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala [hunk 3]
```diff
@@ -192,11 +197,105 @@ class NetworkClient(
...
+  /**
+   * A running server was started with its own `-D` options, and the ones passed to this
+   * invocation would be dropped on the floor. Restart the server so that they take effect.
+   * A server that recorded no options is one that runs without any, so a client that
+   * carries some restarts it too. Completion queries never restart it: a client is not
+   * worth a server to someone pressing tab.
+   */
+  private def restartServerIfSysPropsChanged(promptCompleteUsers: Boolean): Unit =
+    val checked = arguments.forwardsSysProps && !shutdownOnly && !exitOnly && !arguments.bsp &&
+      !promptCompleteUsers && serverAutoStart && serverAutoRestart &&
```
Comment (anatoliykmetyuk) [inline main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: When autostart is disabled, this skips mismatch detection entirely and allows the command to run against a server with stale JVM properties. Could we still detect the mismatch here and fail clearly instead of restarting?

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala [hunk 4]
```diff
@@ -192,11 +197,105 @@ class NetworkClient(
...
+      Try(ClientSocket.portFile(portfile)).toOption match
+        case Some(pf) if pf.sysProps != current =>
+          val dropped = pf.sysProps.diff(current)
+          val added = current.diff(pf.sysProps)
+          console.appendLog(
+            Level.Info,
+            "sbt server is running with different JVM options; restarting it"
+          )
+          if dropped.nonEmpty then
+            console.appendLog(Level.Info, s"dropped: ${dropped.mkString(" ")}")
```
Comment (anatoliykmetyuk) [inline main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: These diagnostics expose complete -D values, which may contain secrets like credentials or tokens. Could we print only the affected property names?

File: protocol/src/main/contraband/portfile.contra [hunk 5]
```diff
@@ -9,6 +9,8 @@ type PortFile {
   uri: String!
   tokenfilePath: String
   tokenfileUri: String
+  ## The -D options the thin client passed to this server when it started it.
+  sysProps: [String] @since("2.0.7")
```
Comment (anatoliykmetyuk) [inline protocol/src/main/contraband/portfile.contra:13]: Could the port file store property names with salted value hashes instead of raw -D strings? That would preserve change detection without persisting possible secrets.

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala [hunk 6]
```diff
@@ -1258,6 +1379,45 @@ object NetworkClient {
...
+    "sbt.supershell",
+  )
+
+  /** The `-D` options that make up the identity of a server started with `sbtArguments`. */
+  private[client] def serverSysProps(sbtArguments: Seq[String]): Seq[String] =
+    sbtArguments
+      .filter(_.startsWith("-D"))
+      .filterNot(a => ignoredSysProps(a.drop(2).takeWhile(_ != '=')))
+      .distinct
+      .sorted
```
Comment (anatoliykmetyuk) [inline main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: Sorting makes property order irrelevant, but order is significant when the same key appears more than once because the JVM uses the last definition. For example, `-Dx=1 -Dx=2` and `-Dx=2 -Dx=1` produce different effective values but normalize to the same sequence here, so the second invocation reuses a server with the wrong value. Could we normalize to each key’s effective final value and add a regression test for reversed duplicate definitions?

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala [hunk 7]
```diff
@@ -431,6 +530,17 @@ class NetworkClient(
             .redirectOutput(nullFile)
             .redirectError(stderrFile)
         processBuilder.environment.put(Terminal.TERMINAL_PROPS, props)
+        if (arguments.forwardsSysProps) {
+          processBuilder.environment.put(
+            NetworkClient.sysPropsEnv,
+            NetworkClient.serverSysProps(arguments.sbtArguments).mkString("\n")
```
Comment (anatoliykmetyuk) [inline main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: Using newline as the separator is ambiguous because a valid -D value may itself contain newlines. Such a property is recorded as multiple entries, so the next identical invocation appears different and restarts the server again. Could this use a lossless encoding such as JSON, with a multiline-value round-trip test?

This is a low-probability issue as users usually pass multiline values via files. However it does impact the correctness of the comparison.

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala (hunk 2 above)
Comment (stasimus) [reply main-command/src/main/scala/sbt/internal/client/NetworkClient.scala:1470]: The client can't see those. The `sbt` script captures the args it passes to sbtn before it prepends `.sbtopts`, `SBT_OPTS` is read the same way, and `.jvmopts` goes into `JAVA_OPTS`. The script re-reads all three when it starts the server, so the server does get them, the client just has nothing to compare against. Closing that needs the launcher script to pass the effective `-D` options down to the client, which I'd rather keep out of this PR. Noted as a limitation in the release note.

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala (hunk 3 above)
Comment (stasimus) [reply main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: Added. The check now runs regardless of autostart/autorestart and warns with the property names when the server is left alone, pointing at `sbt shutdown`. I stopped short of failing: `autorestart=false` is an explicit leave-my-server-alone, and erroring out would break invocations where the options don't matter. Two tests cover it.

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala (hunk 4 above)
Comment (stasimus) [reply main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: Fixed, names only now: `changed: my.prop`. The values are not recorded at all anymore, see the other thread.

File: protocol/src/main/contraband/portfile.contra (hunk 5 above)
Comment (stasimus) [reply protocol/src/main/contraband/portfile.contra:13]: Done. Each option is recorded as `<name>=<salt>:<sha256(salt + option)>` with a fresh salt per entry, so the salt needs no field of its own. The client does the hashing, which also keeps the values out of `SBT_SERVER_SYS_PROPS`, and the server passes that variable on to everything it forks.

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala (hunk 6 above)
Comment (stasimus) [reply main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: Fixed: the options are folded into a name-keyed map that keeps the last definition, so `-Dx=1 -Dx=2` normalizes to `-Dx=2` and the reverse to `-Dx=1`. Test for the reversed pair added.

File: main-command/src/main/scala/sbt/internal/client/NetworkClient.scala (hunk 7 above)
Comment (stasimus) [reply main-command/src/main/scala/sbt/internal/client/NetworkClient.scala]: Fixed: entries are base64 encoded before they are joined with newlines. With the values hashed an entry cannot contain a newline anyway, but the encoding keeps that independent of the digest format. Round trip test with a multiline value added.

**Review by anatoliykmetyuk — APPROVED** [review]
LGTM, thanks!
